fdbserver/storageserver.actor.cpp

```diff
@@ -9446,6 +9449,14 @@ ACTOR Future<Void> fetchKeys(StorageServer* data, AddingShard* shard) {
 			}
 		} // fetchKeys loop.
 
+		if (conductBulkLoad && !SERVER_KNOBS->BULK_LOAD_USE_SST_INGEST &&
+		    data->storage.getKeyValueStore()->supportsSstIngestion()) {
+			// Wait until the load data has been committed
+			wait(data->durableVersion.whenAtLeast(data->storageVersion() + 1));
+			// Compact the range to avoid accumulating TODO compaction
+			wait(data->storage.getKeyValueStore()->compactRange(keys));
```

Comment:
There is the TODO again above.

Why the need for all the compactions? There is a compaction before we ingest, one after we ingest, and then another on the end of the fetchkey loop.  We need all these?  Are compactions even happening? Thanks.

Reply:
My thinking. Before ingest, we do a clearrange, and then a compaction. The ingest goes into the lowest level IIRC (IIRC, to be checked) so the call to compact after the ingest would be a noop.  Ditto on the compaction on the end?

Reply:
If L0, ok.

Reply:
Looks like when a RocksDB ingests a SST file, it adds the data to the "lowest possible level". The level is decided by the current rocksdb internal data layout.
It is possible that SST files (e.g. a ~ b and e ~ f) are added to the bottom level initially. However, the compaction is triggered and the two SST files are merged into one SST (a ~ f). Then, a SST file is ingested (c ~ d). Since this file is conflict to the bottom file (a ~ f), this file cannot be added to the bottom level.

Reply:
SSTs span shards so could have the situation you suggest in the above. We'll probably want to come back here when we start bulkloading into live clusters.